---
title: "Class 18: Investigating cancer genomics datasets"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Connect to the NCI-GDC database

The NCI-GDC (cancer genomic commons) is a major portal for cancer genomics data.

Here we use the bioconductor package **GenomicDataCommons** to talk to this database and download data etc. all from R.

```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

Can we talk to the online database?
```{r}
status()
```

## Querying the GDC from R

The pipe operator can be useful in R and sometimes is more clear to read.

```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

## Plot this overview data

```{r}
x <- cases_by_project$project.project_id
barplot(x$doc_count,  names.arg = x$key, las = 2, col = "lightblue", log = "y")
```




## Variant analysis with R

Find files available for PAAD project that we can then examine variants from

```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
maf.file
```

Read in the MAF file we found

```{r}
vars = read.maf(maf = maf.file, verbose = FALSE)
```

Now make a summary plot of this data

```{r}
plotmafSummary(vars)
```


## Oncoplots

```{r}
library(bio3d)
```




```{r}
#seq <- read.fasta("lecture18_sequences.fa")
#seq
#aln <- seqaln(seq)
```

```{r}
#ide <- conserv(seq$ali, method="identity")
#mutant.sites <- which(ide < 1)
#mutant.sites
```

We are including the gap positions here and this is not what we want. So we need to exclude the gap positions

```{r}
# Exclude gap positions from analysis
#gaps <- gap.inspect(seq)
#gaps$f.inds

#mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

#mutant.sites
```

Find the -8 before and +8 after (i.e. the sequence segment containing all possible 9-mer sequences with a tumor specific amino-acid)

```{r}
# Sequence surrounding each mutant site
#start.position <- mutant.sites -8
#end.position <- mutant.sites +8

```


```{r}
#seq$ali[ 2 , start.position[1]:end.position[1]]
```

```{r}
#mutant.names <- paste0(seq$ali["P53_wt",mutant.sites],
                       #mutant.sites,
                       #seq$ali["P53_mutant",mutant.sites])

#mutant.names
# Blank matrix to store sub-sequences
#store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
#rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
#for(i in 1:length(mutant.sites)) {
 # store.seqs[i,] <- seq$ali["P53_mutant",start.position[i]:end.position[i]]
#}

#store.seqs
```

```{r}
#write.fasta(seq = store.seqs, ids = mutant.names, file = "mutantseqs.fa", gap = FALSE)
```


















